### PR TITLE
Crossbow Cartridges

### PR DESCRIPTION
**Pull request by Dennis-0**
Hold certain items in your offhand to shoot them with a crossbow.

- Bucket of fish (any type): Arrow is replaced by fish and bucket of fish is replaced by a water bucket (if survival mode) -> gives advancement
- Potions (splash/lingering, any type): Arrow is replaced by potion
- Crops (seeds, carrots, potatoes): When the arrow lands on top of a dirt type block, farmland is placed here with this crop
- Torches: Places torch where the arrow hits a block
- Fire charge: shoots blaze fireball (places 1 fire block on aimed block)
- Anvils (decided to keep this one in the datapack, but commented out because the anvil appears late and it seems a little odd)

This could possibly be merged with the roped arrows datapack later.

The advancement is still open to changes

### CODE REVIEW

**Comment by Bluefire-610** [comment]
Great concept!

File: gm4_better_crossbows/data/better_crossbows/functions/shoot/cod.mcfunction [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# @s = player using the crossbow
+
+#summon cod at player's location
+summon minecraft:cod ~ ~1.2 ~ {FromBucket:1,Tags:["gm4_cb_proj"],ActiveEffects:[{Id:10,Amplifier:10,Duration:200,ShowParticles:0b}]}
+
+#replace bucket of cod in player's offhand with water bucket
+execute unless entity @s[gamemode=creative] run replaceitem entity @s[nbt={Inventory:[{Slot:-106b,id:"minecraft:cod_bucket"}]}] weapon.offhand minecraft:water_bucket
```
Comment (Bluefire-610) [inline gm4_better_crossbows/data/better_crossbows/functions/shoot/cod.mcfunction]:
```suggestion
execute unless entity @s[gamemode=creative] run replaceitem entity @s weapon.offhand minecraft:water_bucket
```
This nbt check isn't required as the item was already checked in `check_projectile.mcfunction`

File: gm4_better_crossbows/data/better_crossbows/functions/shoot/lingerpotion.mcfunction [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# @s = player using the crossbow
+
+#summon potion at player's location and copy potion type
+summon minecraft:potion ~ ~1.2 ~ {Tags:["gm4_cb_proj"],Potion:{id:"minecraft:lingering_potion",Count:1b,tag:{Potion:"minecraft:water"}}}
+data modify entity @e[type=minecraft:potion,tag=gm4_cb_proj,distance=..2,limit=1] Potion.tag.Potion set from entity @s Inventory[{Slot:-106b}].tag.Potion
+
+#remove potion from player's offhand
+execute unless entity @s[gamemode=creative] run replaceitem entity @s[nbt={Inventory:[{Slot:-106b,id:"minecraft:lingering_potion"}]}] weapon.offhand minecraft:air
```
Comment (Bluefire-610) [inline gm4_better_crossbows/data/better_crossbows/functions/shoot/lingerpotion.mcfunction]:
```suggestion
execute unless entity @s[gamemode=creative] run replaceitem entity @s weapon.offhand minecraft:air
```
This nbt check isn't required as the item was already checked in `check_projectile.mcfunction`

File: gm4_better_crossbows/data/better_crossbows/functions/shoot/pufferfish.mcfunction [hunk 3]
```diff
@@ -0,0 +1,13 @@
+# @s = player using the crossbow
+
+#summon pufferfish at player's location
+summon minecraft:pufferfish ~ ~1.2 ~ {FromBucket:1,Tags:["gm4_cb_proj"],ActiveEffects:[{Id:10,Amplifier:10,Duration:200,ShowParticles:0b}]}
+
+#replace bucket of pufferfish in player's offhand with water bucket
+execute unless entity @s[gamemode=creative] run replaceitem entity @s[nbt={Inventory:[{Slot:-106b,id:"minecraft:pufferfish_bucket"}]}] weapon.offhand minecraft:water_bucket
```
Comment (Bluefire-610) [inline gm4_better_crossbows/data/better_crossbows/functions/shoot/pufferfish.mcfunction]:
```suggestion
execute unless entity @s[gamemode=creative] run replaceitem entity @s weapon.offhand minecraft:water_bucket
```
This nbt check isn't required as the item was already checked in `check_projectile.mcfunction`

File: gm4_better_crossbows/data/better_crossbows/functions/shoot/salmon.mcfunction [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# @s = player using the crossbow
+
+#summon salmon at player's location
+summon minecraft:salmon ~ ~1.2 ~ {FromBucket:1,Tags:["gm4_cb_proj"],ActiveEffects:[{Id:10,Amplifier:10,Duration:200,ShowParticles:0b}]}
+
+#replace bucket of salmon in player's offhand with water bucket
+execute unless entity @s[gamemode=creative] run replaceitem entity @s[nbt={Inventory:[{Slot:-106b,id:"minecraft:salmon_bucket"}]}] weapon.offhand minecraft:water_bucket
```
Comment (Bluefire-610) [inline gm4_better_crossbows/data/better_crossbows/functions/shoot/salmon.mcfunction]:
```suggestion
execute unless entity @s[gamemode=creative] run replaceitem entity @s weapon.offhand minecraft:water_bucket
```
This nbt check isn't required as the item was already checked in `check_projectile.mcfunction`

File: gm4_better_crossbows/data/better_crossbows/functions/shoot/splashpotion.mcfunction [hunk 5]
```diff
@@ -0,0 +1,11 @@
+# @s = player using the crossbow
+
+#summon potion at player's location and copy potion type
+summon minecraft:potion ~ ~1.2 ~ {Tags:["gm4_cb_proj"],Potion:{id:"minecraft:splash_potion",Count:1b,tag:{Potion:"minecraft:water"}}}
+data modify entity @e[type=minecraft:potion,tag=gm4_cb_proj,distance=..2,limit=1] Potion.tag.Potion set from entity @s Inventory[{Slot:-106b}].tag.Potion
+
+#remove potion from player's offhand
+execute unless entity @s[gamemode=creative] run replaceitem entity @s[nbt={Inventory:[{Slot:-106b,id:"minecraft:splash_potion"}]}] weapon.offhand minecraft:air
```
Comment (Bluefire-610) [inline gm4_better_crossbows/data/better_crossbows/functions/shoot/splashpotion.mcfunction]:
```suggestion
execute unless entity @s[gamemode=creative] run replaceitem entity @s weapon.offhand minecraft:air
```
This nbt check isn't required as the item was already checked in `check_projectile.mcfunction`

File: gm4_better_crossbows/data/better_crossbows/functions/shoot/tropical.mcfunction [hunk 6]
```diff
@@ -0,0 +1,14 @@
+# @s = player using the crossbow
+
+#summon tropical fish at player's location and copy variant
+summon minecraft:tropical_fish ~ ~1.2 ~ {FromBucket:1,Tags:["gm4_cb_proj"],ActiveEffects:[{Id:10,Amplifier:10,Duration:200,ShowParticles:0b}]}
+data modify entity @e[type=minecraft:tropical_fish,tag=gm4_cb_proj,distance=..2,limit=1] Variant set from entity @s Inventory[{Slot:-106b}].tag.BucketVariantTag
+
+#replace bucket of tropical fish in player's offhand with water bucket
+execute unless entity @s[gamemode=creative] run replaceitem entity @s[nbt={Inventory:[{Slot:-106b,id:"minecraft:tropical_fish_bucket"}]}] weapon.offhand minecraft:water_bucket
```
Comment (Bluefire-610) [inline gm4_better_crossbows/data/better_crossbows/functions/shoot/tropical.mcfunction]:
```suggestion
execute unless entity @s[gamemode=creative] run replaceitem entity @s weapon.offhand minecraft:water_bucket
```
This nbt check isn't required as the item was already checked in `check_projectile.mcfunction`

File: gm4_better_crossbows/data/better_crossbows/functions/place_torch.mcfunction [hunk 7]
```diff
@@ -0,0 +1,19 @@
+# @s = @e[type=minecraft:arrow,tag=gm4_cb_torch,nbt={inGround:1b}]
+# at @s
+
+#place torch
+execute if block ~ ~-.05 ~ #better_crossbows:torch_placeable run setblock ~ ~ ~ minecraft:torch
```
Comment (Bluefire-610) [inline gm4_better_crossbows/data/better_crossbows/functions/place_torch.mcfunction]: Move this line to line 10 to make the "default" standing.

Comment (Dennis-0) [reply gm4_better_crossbows/data/better_crossbows/functions/place_torch.mcfunction]: it's checking for 0.05 blocks below or to the side, so "default" is only when it's exactly in the at the edge of the block

**Pull request closed without merging**